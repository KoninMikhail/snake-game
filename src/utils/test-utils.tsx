import React, {PropsWithChildren} from 'react';
import type {RenderOptions} from '@testing-library/react';
import {render} from '@testing-library/react';
import type {PreloadedState} from '@reduxjs/toolkit';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import type {AppStore, RootState} from '@store/store';
import appSlice from '@store/slices/app.slice';
import soundSlice from '@store/slices/sound.slice';
import maxScoreSlice from '@store/slices/maxScore.slice';
import gameDataSlice from '@store/slices/game.slice';
import {ImagesData} from '@data/imagesData';
import {mockImagesData, mockSoundData} from '@/setupTests';
import SoundsProvider from '@context/SoundsContext';
import ImagesProvider from '@context/ImagesContext';
import {SoundsData} from '@data/soundsData';
import {DefaultTheme, ThemeProvider} from 'styled-components';
import {theme as appTheme} from '@styles/theme';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
    images?: ImagesData;
    sounds?: SoundsData;
    theme?: DefaultTheme;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                app: appSlice,
                sounds: soundSlice,
                maxScore: maxScoreSlice,
                game: gameDataSlice,
            },
            preloadedState,
        }),
        images = mockImagesData,
        sounds = mockSoundData,
        theme = appTheme,
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({children}: PropsWithChildren): JSX.Element {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <SoundsProvider sounds={sounds}>
                        <ImagesProvider images={images}>{children}</ImagesProvider>
                    </SoundsProvider>
                </ThemeProvider>
            </Provider>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
