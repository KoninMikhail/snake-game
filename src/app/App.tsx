import { useEffect } from 'react';
import { useEffectOnce, useUpdateEffect } from 'usehooks-ts';
import { Outlet, ScrollRestoration, useLocation } from 'react-router-dom';
import { FontStyles } from '@styles/fonts-face';
import { GlobalStyles } from '@styles/globals';
import { appActions } from '@store/slices/app.slice';
import { startAppListening } from '@store/store';
import useWindowOrientation from '@hooks/useWindowOrientation/useWindowOrientation';
import useDeviceScreenType from '@hooks/useDeviceScreenType/useDeviceScreenType';
import useActionCreators from '@hooks/useActionCreators/useActionCreators';
import { setupAppListeners, setupGameListeners } from '@store/middlewares/setupListeners';
import { Loader } from '@ui/atoms/loaders/Loader/Loader';
import { isTouchDevice } from '@helpers/validators/isTouchDevice';

/**
 * App
 */
const App = () => {
    const actions = useActionCreators(appActions);
    const location = useLocation();

    /** Detect Screen type
     * ======================== */
    const screen = useDeviceScreenType();
    useUpdateEffect(() => {
        actions.setDeviceScreenType({ screenType: screen.type });
    }, [screen]);

    /** Detect Orientation
     * ======================== */
    const { orientation } = useWindowOrientation();
    useEffect(() => {
        actions.setDeviceOrientation({ orientation });
    }, [actions, orientation]);

    /** Detect touch
     * ======================== */
    useEffectOnce(() => {
        isTouchDevice() && actions.setDeviceTouchState({ touch: true });
    });

    /** Detect Change Route
     * ======================== */
    useEffect(() => {
        actions.setRoute({ route: location.pathname });
    }, [actions, location.pathname]);

    /** Middlewares
     * ======================== */
    useEffectOnce(() => {
        setupAppListeners(startAppListening);
        return () => setupGameListeners(startAppListening)();
    });
    return (
        <>
            <GlobalStyles />
            <FontStyles />
            <Loader />
            <Outlet />
            <ScrollRestoration />
        </>
    );
};

export default App;
