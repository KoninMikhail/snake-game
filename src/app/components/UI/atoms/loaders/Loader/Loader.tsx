import {LoaderOverlay} from '@ui/atoms/loaders/Loader/style/StyledLoader';
import useAppSelector from '@hooks/useAppSelector/useAppSelector';
import {appPageStateSelector} from '@store/slices/app.slice';

/**
 * @name Loader
 * @author Mikhail Konin <dev.konin@gmail.com>
 *
 * Use it for showing overlay on page
 */
export const Loader = () => {
    const state = useAppSelector(appPageStateSelector);
    const isLoading = state === 'loading';
    return <LoaderOverlay show={isLoading} />;
};
