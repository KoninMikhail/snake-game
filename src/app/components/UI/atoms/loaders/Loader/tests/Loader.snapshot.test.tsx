import {AppGlobalState, PageState} from '@store/slices/app.slice';
import {renderWithProviders} from '@utils/test-utils';
import {Loader} from '@ui/atoms/loaders/Loader/Loader';

/**
 * Test
 */
describe('Loader', () => {
    const testId = 'overlayEl';

    it('correctly render when hidden', () => {
        const initialState = {
            preloadedState: { app: { pageState: 'loading' as PageState } as AppGlobalState },
        };
        const { container } = renderWithProviders(<Loader data-testid={testId} />, initialState);
        expect(container).toMatchSnapshot();
    });

    it('correctly render when showed', () => {
        const initialState = {
            preloadedState: { app: { pageState: 'ready' as PageState } as AppGlobalState },
        };
        const { container } = renderWithProviders(<Loader data-testid={testId} />, initialState);
        expect(container).toMatchSnapshot();
    });
});
