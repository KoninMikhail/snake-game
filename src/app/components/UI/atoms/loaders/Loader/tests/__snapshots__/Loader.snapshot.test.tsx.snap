// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader correctly render when hidden 1`] = `
.c0 {
  display: grid;
  gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: black;
  opacity: 1;
  visibility: visible;
  -webkit-transition: visibility 0.3s,opacity 0.9s linear;
  transition: visibility 0.3s,opacity 0.9s linear;
  z-index: 9100;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Loader correctly render when showed 1`] = `
.c0 {
  display: grid;
  gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: black;
  opacity: 0;
  visibility: hidden;
  -webkit-transition: visibility 0.3s,opacity 0.9s linear;
  transition: visibility 0.3s,opacity 0.9s linear;
  z-index: 9100;
}

<div>
  <div
    class="c0"
  />
</div>
`;
